src/shims/io.rs

```diff
@@ -133,28 +154,63 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         let count = this.read_scalar(count_op)?.to_usize(&*this.tcx)?;
 
         // Remove the file handle to avoid borrowing issues
-        this.remove_handle_and(
-            fd,
-            |mut handle, this| {
-                let bytes = handle
-                    .file
-                    .read(this.memory_mut().get_mut(buf.alloc_id)?.get_bytes_mut(
-                        tcx,
-                        buf,
-                        Size::from_bytes(count),
-                    )?)
-                    .map(|bytes| bytes as i64);
-                // Reinsert the file handle
-                this.machine.file_handler.handles.insert(fd, handle);
-                this.consume_result(bytes)
-            },
-        )
+        this.remove_handle_and(fd, |mut handle, this| {
+            // Don't use `?` to avoid returning before reinserting the handle
+
+            let bytes = this.memory_mut().get_mut(buf.alloc_id).and_then(|alloc| {
+                if count != 0 {
+                    alloc
+                        .get_bytes_mut(tcx, buf, Size::from_bytes(count))
+                        .map(|buffer| handle.file.read(buffer))
+                } else {
+                    // Using `Allocation::get_bytes_mut` requires to actually put data into
+                    // the allocation even if we take 0 bytes of it. In that case we use an
+                    // empty buffer instead to avoid dangling pointers.
+                    let mut empty = Vec::new();
+                    let result = Ok(handle.file.read(empty.as_mut_slice()));
+                    assert!(empty.is_empty());
+                    result
+                }
+            });
+            // Reinsert the file handle
+            this.machine.file_handler.handles.insert(fd, handle);
+            this.consume_result(bytes?.map(|bytes| bytes as i64))
+        })
+    }
+
+    fn write(
+        &mut self,
+        fd_op: OpTy<'tcx, Tag>,
+        buf_op: OpTy<'tcx, Tag>,
+        count_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i64> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`write` not available when isolation is enabled")
+        }
+
+        let tcx = &{ this.tcx.tcx };
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let buf = this.force_ptr(this.read_scalar(buf_op)?.not_undef()?)?;
+        let count = this.read_scalar(count_op)?.to_usize(&*this.tcx)?;
+
+        this.remove_handle_and(fd, |mut handle, this| {
+            let bytes = this.memory().get(buf.alloc_id).and_then(|alloc| {
+                alloc
+                    .get_bytes(tcx, buf, Size::from_bytes(count))
```

Comment:
You can use `?` inside the closure without any problems if that makes the code simpler